server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java

```diff
@@ -22,138 +24,425 @@
+     *                               an empty result
+     */
+    public record WildcardOptions(
+        boolean matchOpen,
+        boolean matchClosed,
+        boolean includeHidden,
+        boolean resolveAliases,
+        boolean allowEmptyExpressions
+    ) implements ToXContentFragment {
+
+        public static final WildcardOptions DEFAULT = WildcardOptions.newBuilder().build();
+
+        public static WildcardOptions parseParameters(Object expandWildcards, Object allowNoIndices, WildcardOptions defaultOptions) {
+            if (expandWildcards == null && allowNoIndices == null) {
+                return defaultOptions;
             }
+            WildcardOptions.Builder builder = defaultOptions == null ? new Builder() : new Builder(defaultOptions);
+            if (expandWildcards != null) {
+                builder.none();
+                builder.expandStates(nodeStringArrayValue(expandWildcards));
+            }
+
+            if (allowNoIndices != null) {
+                builder.allowEmptyExpressions(nodeBooleanValue(allowNoIndices, "allow_no_indices"));
+            }
+            return builder.build();
+        }
 
-            return states;
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            return toXContent(builder, false);
         }
 
-        public static XContentBuilder toXContent(EnumSet<WildcardStates> states, XContentBuilder builder) throws IOException {
-            if (states.isEmpty()) {
-                builder.field("expand_wildcards", "none");
-            } else if (states.containsAll(EnumSet.allOf(WildcardStates.class))) {
-                builder.field("expand_wildcards", "all");
+        /**
+         * This converter to XContent only includes the fields a user can interact, internal options like the resolveAliases
+         * are not added.
+         * @param wildcardStatesAsUserInput, some parts of the code expect the serialization of the expand_wildcards field
+         *                                   to be a comma separated string that matches the allowed user input, this includes
+         *                                   all the states along with the values 'all' and 'none'.
+         */
+        public XContentBuilder toXContent(XContentBuilder builder, boolean wildcardStatesAsUserInput) throws IOException {
+            List<String> legacyStates = new ArrayList<>(3);
+            if (matchOpen()) {
+                legacyStates.add("open");
+            }
+            if (matchClosed()) {
+                legacyStates.add("closed");
+            }
+            if (includeHidden()) {
+                legacyStates.add("hidden");
+            }
+            if (wildcardStatesAsUserInput) {
+                if (legacyStates.isEmpty()) {
+                    builder.field("expand_wildcards", "none");
```

Comment:
Same comment here about making `"expand_wildcards"` a public static string (not that we ever intend to rename it, but we could if we wanted to)